# NYC Taxi fare problem - Assignment 1
***
> ##### __Author__: _Vanshaj Lokwani_
> ##### __SBU ID__: _112025869_

In [2]:
# Importing the tools required
import pandas as pd
import numpy as np
import sklearn as skl
from matplotlib import pyplot as plt
import seaborn as sns

The dataset contains 55M rows and trying to load all of them at once causes the system to hang and intermittently crash. For the initial cleaning and testing phase, using a subset of the data. Once we have an idea of the actual cleaning effort required, we can iteratively perform that on the whole dataset for further processing. 

__For the time being, using first 10M rows as our subset. There is no reason behind this and the number was selected by intuition. Choosing 10M rows gives us enough liberty with the data to identify patterns and the data is small enough to not overload the RAM during pre-processing__

In [3]:
# Getting the dataset
NUM_ROWS = 10**7 # 10M ROWS
train_df = pd.read_csv('../dataset/train.csv', nrows=NUM_ROWS)

Displaying the first 5 rows from the dataframe to get a rough idea of the DF looks like and if it's loaded correctly. 

In [4]:
display(train_df.head())

key  fare_amount          pickup_datetime  \
0    2009-06-15 17:26:21.0000001          4.5  2009-06-15 17:26:21 UTC   
1    2010-01-05 16:52:16.0000002         16.9  2010-01-05 16:52:16 UTC   
2   2011-08-18 00:35:00.00000049          5.7  2011-08-18 00:35:00 UTC   
3    2012-04-21 04:30:42.0000001          7.7  2012-04-21 04:30:42 UTC   
4  2010-03-09 07:51:00.000000135          5.3  2010-03-09 07:51:00 UTC   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.844311        40.721319         -73.841610         40.712278   
1        -74.016048        40.711303         -73.979268         40.782004   
2        -73.982738        40.761270         -73.991242         40.750562   
3        -73.987130        40.733143         -73.991567         40.758092   
4        -73.968095        40.768008         -73.956655         40.783762   

   passenger_count  
0                1  
1                1  
2                2  
3                1  
4                1

In [5]:
# Few steps to make our life easier in the long run.
# set the new index as the 'key' coumn. The dataset specifies that the key is unique overall
# the it does not make sense to maintain two different indexes. 
train_df.set_index('key', inplace=True, drop=True) 

#getting the columns in a list. Might be helpful later on. 
columns = list(train_df.columns)
display(columns)

['fare_amount',
 'pickup_datetime',
 'pickup_longitude',
 'pickup_latitude',
 'dropoff_longitude',
 'dropoff_latitude',
 'passenger_count']

## Question 1: 
##### ** Take a look at the training data. There may be anomalies in the data that you may need to factor in before you start on the other tasks. Clean the data first to handle these issues. Explain what you did to clean the data (in bulleted form). (10 pt) **


##### Let's print a summary of the data using the .describe() api provided in pandas. This may help us detect any anomalies that might exist in the system.

But before we get into that, we need to supress scientific notation in pandas. The easiest way to do this is to call 'set_option'. We'll set the precision for the number to be accurate upto 3 decimal places. More information on how to do that can be found at the stackoverflow link: https://stackoverflow.com/questions/21137150/format-suppress-scientific-notation-from-python-pandas-aggregation-results

In [6]:
pd.set_option('display.float_format', lambda x: '%.3f' % x)
display(train_df.describe())

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count 10000000.000      10000000.000     10000000.000        9999931.000   
mean        11.339           -72.508           39.919            -72.509   
std          9.800            12.994            9.323             12.875   
min       -107.750         -3439.245        -3492.264          -3426.601   
25%          6.000           -73.992           40.735            -73.991   
50%          8.500           -73.982           40.753            -73.980   
75%         12.500           -73.967           40.767            -73.964   
max       1273.310          3457.626         3344.459           3457.622   

       dropoff_latitude  passenger_count  
count       9999931.000     10000000.000  
mean             39.919            1.685  
std               9.237            1.323  
min           -3488.080            0.000  
25%              40.734            1.000  
50%              40.753            1.000  
75%              40.768            2.000  
max            3351.403          208.000

#### We can already see that there are clearly some issues with the data. Let's pick the columns one by one.

__ColumnName__: *pickup_longitude, pickup_latitude, dropoff_longitude & dropoff_latitude* Possible Issues:

* dropoff_longitude and dropoff_latitude count is lesser compared to other columns. This signifies the presence of missing values or NaN values in the two columns. We would have to take care of these depending how many values are missing.

In [7]:
missing_long_indexes = train_df[train_df.dropoff_longitude.isnull().values].dropoff_longitude.index
missing_lat_indexes = train_df[train_df.dropoff_latitude.isnull().values].dropoff_latitude.index
missing_values_list = set(missing_long_indexes).union(set(missing_lat_indexes))
print("Number of rows with missing indexes: {}".format(len(missing_values_list)))

Number of rows with missing indexes: 69


the number seems small enough that these rows can be dropped without worrying about loosing important data. 69 in 10M is not a very significant number.  

In [8]:
train_df.drop(missing_values_list, inplace=True)

In [9]:
# describe api should return equal counts for all columns with no na values
display(train_df.describe())
display("Any null values left in the two columns: %s" % train_df[['dropoff_longitude', 'dropoff_latitude']].isnull().values.any())

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9999931.000       9999931.000      9999931.000        9999931.000   
mean        11.338           -72.508           39.919            -72.509   
std          9.800            12.994            9.323             12.875   
min       -107.750         -3439.245        -3492.264          -3426.601   
25%          6.000           -73.992           40.735            -73.991   
50%          8.500           -73.982           40.753            -73.980   
75%         12.500           -73.967           40.767            -73.964   
max       1273.310          3457.626         3344.459           3457.622   

       dropoff_latitude  passenger_count  
count       9999931.000      9999931.000  
mean             39.919            1.685  
std               9.237            1.323  
min           -3488.080            0.000  
25%              40.734            1.000  
50%              40.753            1.000  
75%              40.768            2.000  
max            3351.403          208.000

'Any null values left in the two columns: False'

* The range of latitude values is -90 to 90 and the range of longitude values is -180 to 180. From the data summary, it's clear that the data does not follow the range. Calculating the number of points which are outside the range: 

In [10]:
# Out of Range = or_
or_pickup_longitude = set(train_df[(train_df.pickup_longitude < -180.0) | (train_df.pickup_longitude > 180.0)].index)
or_pickup_latitude = set(train_df[(train_df.pickup_latitude < -90.0) | (train_df.pickup_latitude > 90.0)].index)
or_dropoff_longitude = set(train_df[(train_df.dropoff_longitude < -180.0) | (train_df.dropoff_longitude > 180.0)].index)
or_dropoff_latitude = set(train_df[(train_df.dropoff_latitude < -90.0) | (train_df.dropoff_latitude > 90.0)].index)
or_indexes = or_pickup_longitude.union(or_pickup_latitude).union(or_dropoff_latitude).union(or_dropoff_longitude)
display("Number of rows with outside range values: %s" % len(or_indexes))

'Number of rows with outside range values: 476'

In [11]:
train_df.loc[list(or_indexes)].describe()

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count      476.000           476.000          476.000            476.000   
mean        10.737          -316.692          116.937           -258.439   
std          8.741          1090.612         1017.041           1075.646   
min          2.500         -3439.245        -3492.264          -3426.601   
25%          5.700          -736.367           40.721           -736.237   
50%          8.100           -73.993           40.757            -73.990   
75%         11.700           -73.960          288.889            -73.953   
max         57.330          3457.626         3344.459           3457.622   

       dropoff_latitude  passenger_count  
count           476.000          476.000  
mean             63.743            1.901  
std            1004.023            2.719  
min           -3488.080            1.000  
25%              40.704            1.000  
50%              40.756            1.000  
75%              40.811            2.000  
max            3351.403           51.000

* 476 in ~10M is a very small percentage of datapoints. We can look into the datapoints individually later and try to figure out if there is any way we can fix these outliers but for now dropping the points.

In [12]:
train_df.drop(or_indexes, inplace=True)

In [13]:
display(train_df.describe())

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9999455.000       9999455.000      9999455.000        9999455.000   
mean        11.339           -72.496           39.916            -72.500   
std          9.800            10.465            6.123             10.449   
min       -107.750          -168.604          -74.824           -173.342   
25%          6.000           -73.992           40.735            -73.991   
50%          8.500           -73.982           40.753            -73.980   
75%         12.500           -73.967           40.767            -73.964   
max       1273.310           154.101           89.742            154.101   

       dropoff_latitude  passenger_count  
count       9999455.000      9999455.000  
mean             39.918            1.685  
std               6.117            1.323  
min             -74.193            0.000  
25%              40.734            1.000  
50%              40.753            1.000  
75%              40.768            2.000  
max              81.555          208.000

* One more thing to note here, is that the data is the fares charged by NYC taxis. If the data is only from NYC taxis, then some of the of longitude/latitude datapoints will become outliers as these points are way beyond the city ( some even outside the United States). A simple search returns that the __range of latitudes and longitudes for NYC__ lie from:

_North Latitude: 40.917577 South Latitude: 40.477399 East Longitude: -73.700272 West Longitude: -74.259090_. 

**Source: https://www.mapdevelopers.com/geocode_bounding_box.php**

In [18]:
# Fetching all points in this range to get a better idea of the sample dataset. 

nyc_coordinates = train_df[(train_df['pickup_longitude'].between(-74.259090, -73.700272)) &
                             (train_df['pickup_latitude'].between(40.477399, 40.917577)) &
                             (train_df['dropoff_longitude'].between(-74.259090, -73.700272)) &
                             (train_df['dropoff_latitude'].between(40.477399, 40.917577))]

display(nyc_coordinates.describe(), nyc_coordinates.head())
nyc_coor_index = nyc_coordinates.index

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9776221.000       9776221.000      9776221.000        9776221.000   
mean        11.279           -73.976           40.751            -73.975   
std          9.460             0.034            0.027              0.034   
min       -107.750           -74.259           40.478            -74.259   
25%          6.000           -73.992           40.737            -73.992   
50%          8.500           -73.982           40.753            -73.981   
75%         12.500           -73.968           40.768            -73.965   
max        952.000           -73.700           40.917            -73.700   

       dropoff_latitude  passenger_count  
count       9776221.000      9776221.000  
mean             40.751            1.685  
std               0.031            1.314  
min              40.478            0.000  
25%              40.736            1.000  
50%              40.754            1.000  
75%              40.768            2.000  
max              40.918          208.000

fare_amount          pickup_datetime  \
key                                                                   
2009-06-15 17:26:21.0000001          4.500  2009-06-15 17:26:21 UTC   
2010-01-05 16:52:16.0000002         16.900  2010-01-05 16:52:16 UTC   
2011-08-18 00:35:00.00000049         5.700  2011-08-18 00:35:00 UTC   
2012-04-21 04:30:42.0000001          7.700  2012-04-21 04:30:42 UTC   
2010-03-09 07:51:00.000000135        5.300  2010-03-09 07:51:00 UTC   

                               pickup_longitude  pickup_latitude  \
key                                                                
2009-06-15 17:26:21.0000001             -73.844           40.721   
2010-01-05 16:52:16.0000002             -74.016           40.711   
2011-08-18 00:35:00.00000049            -73.983           40.761   
2012-04-21 04:30:42.0000001             -73.987           40.733   
2010-03-09 07:51:00.000000135           -73.968           40.768   

                               dropoff_longitude  dropoff_latitude  \
key                                                                  
2009-06-15 17:26:21.0000001              -73.842            40.712   
2010-01-05 16:52:16.0000002              -73.979            40.782   
2011-08-18 00:35:00.00000049             -73.991            40.751   
2012-04-21 04:30:42.0000001              -73.992            40.758   
2010-03-09 07:51:00.000000135            -73.957            40.784   

                               passenger_count  
key                                             
2009-06-15 17:26:21.0000001                  1  
2010-01-05 16:52:16.0000002                  1  
2011-08-18 00:35:00.00000049                 2  
2012-04-21 04:30:42.0000001                  1  
2010-03-09 07:51:00.000000135                1

~200,000 datapoints where dropped with geofencing the dataset. Although, this number is still small in front of ~9.8M datapoints availabel for prediction, but 200,000 in itself is a huge number and these many data points cannot be considered as outliers. Let's explore the dataset a little bit more to actually try and figure out if there is any thing from the remaining dataset that we can use. 

In [20]:
train_df.drop(nyc_coor_index).describe(np.linspace(0,1,101))

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count   223234.000        223234.000       223234.000         223234.000   
mean        13.966            -7.710            3.341             -7.925   
std         19.383            24.747           17.644             25.003   
min       -100.000          -168.604          -74.824           -173.342   
0%        -100.000          -168.604          -74.824           -173.342   
1%           2.500           -74.286          -73.984            -74.291   
2%           2.500           -74.003          -73.950            -74.005   
3%           3.000           -73.991            0.000            -73.991   
4%           3.500           -73.983            0.000            -73.983   
5%           3.700           -73.975            0.000            -73.976   
6%           4.000           -73.963            0.000            -73.964   
7.0%         4.100           -73.934            0.000            -73.948   
8%           4.100           -73.801            0.000            -73.858   
9%           4.500           -73.761            0.000            -73.694   
10%          4.500           -73.137            0.000            -73.573   
11%          4.500           -73.137            0.000            -73.137   
12%          4.500            -0.041            0.000             -7.583   
13%          4.900             0.000            0.000              0.000   
14.0%        4.900             0.000            0.000              0.000   
15%          4.900             0.000            0.000              0.000   
16%          5.000             0.000            0.000              0.000   
17%          5.000             0.000            0.000              0.000   
18%          5.300             0.000            0.000              0.000   
19%          5.300             0.000            0.000              0.000   
20%          5.500             0.000            0.000              0.000   
21%          5.500             0.000            0.000              0.000   
22%          5.700             0.000            0.000              0.000   
23%          5.700             0.000            0.000              0.000   
24%          5.700             0.000            0.000              0.000   
25%          6.000             0.000            0.000              0.000   
...            ...               ...              ...                ...   
72%         12.500             0.000            0.000              0.000   
73%         12.900             0.000            0.000              0.000   
74%         13.000             0.000            0.000              0.000   
75%         13.500             0.000            0.000              0.000   
76%         13.700             0.000            0.000              0.000   
77%         14.100             0.000            0.000              0.000   
78%         14.500             0.000            0.000              0.000   
79%         15.000             0.000            0.000              0.000   
80%         15.500             0.000            0.000              0.000   
81%         16.000             0.000            0.000              0.000   
82%         16.500             0.000            0.000              0.000   
83%         17.300             0.000            0.000              0.000   
84%         18.000             0.000            0.000              0.000   
85%         19.000             0.000            0.000              0.000   
86%         20.000             0.000            0.000              0.000   
87%         21.000             0.000            0.000              0.000   
88%         22.900             0.000           39.604              0.000   
89%         24.900             0.000           40.591              0.000   
90%         27.000             0.000           40.647              0.000   
91%         29.700             0.000           40.719              0.000   
92%         32.500             0.000           40.738  

Looks like most of the values are 0.000 which can be safely ignored ( assuming people are not travelling to the middle of the atlantic ocean). 
some of the coordinates are very close to the range we originally took. These can be included in our calculation considering there might be a small error in calculating the city boundries. Ignoring the remaining latitudes and longitudes, our final range becomes: 
_North Latitude: 41.366000 South Latitude: 40.477399 East Longitude: -73.700272 West Longitude: -74.286090_

In [22]:
# Fetching all points in this range to get a better idea of the sample dataset. 

train_df = train_df[(train_df['pickup_longitude'].between(-74.286090, -73.700272)) &
                             (train_df['pickup_latitude'].between(40.477399, 41.366000)) &
                             (train_df['dropoff_longitude'].between(-74.286090, -73.700272)) &
                             (train_df['dropoff_latitude'].between(40.477399, 41.366000))]
display(train_df.head());display(train_df.describe())

fare_amount          pickup_datetime  \
key                                                                   
2009-06-15 17:26:21.0000001          4.500  2009-06-15 17:26:21 UTC   
2010-01-05 16:52:16.0000002         16.900  2010-01-05 16:52:16 UTC   
2011-08-18 00:35:00.00000049         5.700  2011-08-18 00:35:00 UTC   
2012-04-21 04:30:42.0000001          7.700  2012-04-21 04:30:42 UTC   
2010-03-09 07:51:00.000000135        5.300  2010-03-09 07:51:00 UTC   

                               pickup_longitude  pickup_latitude  \
key                                                                
2009-06-15 17:26:21.0000001             -73.844           40.721   
2010-01-05 16:52:16.0000002             -74.016           40.711   
2011-08-18 00:35:00.00000049            -73.983           40.761   
2012-04-21 04:30:42.0000001             -73.987           40.733   
2010-03-09 07:51:00.000000135           -73.968           40.768   

                               dropoff_longitude  dropoff_latitude  \
key                                                                  
2009-06-15 17:26:21.0000001              -73.842            40.712   
2010-01-05 16:52:16.0000002              -73.979            40.782   
2011-08-18 00:35:00.00000049             -73.991            40.751   
2012-04-21 04:30:42.0000001              -73.992            40.758   
2010-03-09 07:51:00.000000135            -73.957            40.784   

                               passenger_count  
key                                             
2009-06-15 17:26:21.0000001                  1  
2010-01-05 16:52:16.0000002                  1  
2011-08-18 00:35:00.00000049                 2  
2012-04-21 04:30:42.0000001                  1  
2010-03-09 07:51:00.000000135                1

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9779260.000       9779260.000      9779260.000        9779260.000   
mean        11.294           -73.976           40.751            -73.975   
std          9.538             0.035            0.027              0.034   
min       -107.750           -74.285           40.478            -74.286   
25%          6.000           -73.992           40.737            -73.992   
50%          8.500           -73.982           40.753            -73.981   
75%         12.500           -73.968           40.768            -73.965   
max        952.000           -73.700           41.361            -73.700   

       dropoff_latitude  passenger_count  
count       9779260.000      9779260.000  
mean             40.751            1.685  
std               0.031            1.314  
min              40.478            0.000  
25%              40.736            1.000  
50%              40.754            1.000  
75%              40.768            2.000  
max              41.358          208.000


__ColumnName__: *fare_amount*
Possible Issues: 
* fare_amount is negative which should not be possible. Assumption: there was an issue with data entry or fare calculation. looking into the dataset to fetch all the rows where fare_amount is negetive.

In [23]:
neg_fare = train_df[train_df.fare_amount < 0]
display(neg_fare.describe())

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count      380.000           380.000          380.000            380.000   
mean        -8.869           -73.966           40.749            -73.964   
std         14.197             0.052            0.036              0.054   
min       -107.750           -74.230           40.583            -74.230   
25%         -6.100           -73.991           40.734            -73.990   
50%         -3.500           -73.979           40.755            -73.977   
75%         -2.500           -73.959           40.766            -73.954   
max         -0.010           -73.727           40.919            -73.727   

       dropoff_latitude  passenger_count  
count           380.000          380.000  
mean             40.747            1.824  
std               0.041            1.387  
min              40.588            0.000  
25%              40.732            1.000  
50%              40.753            1.000  
75%              40.766            2.000  
max              40.919            6.000

Looking at the number of data points that have negative fare value is just 420 in ~10M. We can safely drop these values from the dataset as well.

In [24]:
train_df.drop(neg_fare.index, inplace=True)
display(train_df.head())
display(train_df.describe())

fare_amount          pickup_datetime  \
key                                                                   
2009-06-15 17:26:21.0000001          4.500  2009-06-15 17:26:21 UTC   
2010-01-05 16:52:16.0000002         16.900  2010-01-05 16:52:16 UTC   
2011-08-18 00:35:00.00000049         5.700  2011-08-18 00:35:00 UTC   
2012-04-21 04:30:42.0000001          7.700  2012-04-21 04:30:42 UTC   
2010-03-09 07:51:00.000000135        5.300  2010-03-09 07:51:00 UTC   

                               pickup_longitude  pickup_latitude  \
key                                                                
2009-06-15 17:26:21.0000001             -73.844           40.721   
2010-01-05 16:52:16.0000002             -74.016           40.711   
2011-08-18 00:35:00.00000049            -73.983           40.761   
2012-04-21 04:30:42.0000001             -73.987           40.733   
2010-03-09 07:51:00.000000135           -73.968           40.768   

                               dropoff_longitude  dropoff_latitude  \
key                                                                  
2009-06-15 17:26:21.0000001              -73.842            40.712   
2010-01-05 16:52:16.0000002              -73.979            40.782   
2011-08-18 00:35:00.00000049             -73.991            40.751   
2012-04-21 04:30:42.0000001              -73.992            40.758   
2010-03-09 07:51:00.000000135            -73.957            40.784   

                               passenger_count  
key                                             
2009-06-15 17:26:21.0000001                  1  
2010-01-05 16:52:16.0000002                  1  
2011-08-18 00:35:00.00000049                 2  
2012-04-21 04:30:42.0000001                  1  
2010-03-09 07:51:00.000000135                1

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9778880.000       9778880.000      9778880.000        9778880.000   
mean        11.295           -73.976           40.751            -73.975   
std          9.537             0.035            0.027              0.034   
min          0.000           -74.285           40.478            -74.286   
25%          6.000           -73.992           40.737            -73.992   
50%          8.500           -73.982           40.753            -73.981   
75%         12.500           -73.968           40.768            -73.965   
max        952.000           -73.700           41.361            -73.700   

       dropoff_latitude  passenger_count  
count       9778880.000      9778880.000  
mean             40.751            1.685  
std               0.031            1.314  
min              40.478            0.000  
25%              40.736            1.000  
50%              40.754            1.000  
75%              40.768            2.000  
max              41.358          208.000

__ColumnName__: *passenger_count* Possible Issues:
* The maximum value of passenger_count reaches 208, which does not seem like it would be possible for a taxi. the maximum number of people allowed in a NYC taxi is 5 + a child below the age of 7. Let's take the case of a child in a taxi as a special case that we will deal with at a later point in time. Lets look at data where passenger count > 5.

In [25]:
train_df[train_df.passenger_count > 5.0].groupby('passenger_count').count()

fare_amount  pickup_datetime  pickup_longitude  \
passenger_count                                                   
6                     207357           207357            207357   
7                          2                2                 2   
9                          3                3                 3   
129                        1                1                 1   
208                        3                3                 3   

                 pickup_latitude  dropoff_longitude  dropoff_latitude  
passenger_count                                                        
6                         207357             207357            207357  
7                              2                  2                 2  
9                              3                  3                 3  
129                            1                  1                 1  
208                            3                  3                 3

<p>The number of datapoints above 6 passengers is few in number. These can be considered as outliers and dropped safely. There is no point trying to fix the passenger counts because the change caused by these number will not be significant</p>

In [26]:
train_df.drop(train_df[train_df.passenger_count > 6].index, inplace=True)
display(train_df.head())

fare_amount          pickup_datetime  \
key                                                                   
2009-06-15 17:26:21.0000001          4.500  2009-06-15 17:26:21 UTC   
2010-01-05 16:52:16.0000002         16.900  2010-01-05 16:52:16 UTC   
2011-08-18 00:35:00.00000049         5.700  2011-08-18 00:35:00 UTC   
2012-04-21 04:30:42.0000001          7.700  2012-04-21 04:30:42 UTC   
2010-03-09 07:51:00.000000135        5.300  2010-03-09 07:51:00 UTC   

                               pickup_longitude  pickup_latitude  \
key                                                                
2009-06-15 17:26:21.0000001             -73.844           40.721   
2010-01-05 16:52:16.0000002             -74.016           40.711   
2011-08-18 00:35:00.00000049            -73.983           40.761   
2012-04-21 04:30:42.0000001             -73.987           40.733   
2010-03-09 07:51:00.000000135           -73.968           40.768   

                               dropoff_longitude  dropoff_latitude  \
key                                                                  
2009-06-15 17:26:21.0000001              -73.842            40.712   
2010-01-05 16:52:16.0000002              -73.979            40.782   
2011-08-18 00:35:00.00000049             -73.991            40.751   
2012-04-21 04:30:42.0000001              -73.992            40.758   
2010-03-09 07:51:00.000000135            -73.957            40.784   

                               passenger_count  
key                                             
2009-06-15 17:26:21.0000001                  1  
2010-01-05 16:52:16.0000002                  1  
2011-08-18 00:35:00.00000049                 2  
2012-04-21 04:30:42.0000001                  1  
2010-03-09 07:51:00.000000135                1

In [27]:
display(train_df.describe())

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9778871.000       9778871.000      9778871.000        9778871.000   
mean        11.294           -73.976           40.751            -73.975   
std          9.537             0.035            0.027              0.034   
min          0.000           -74.285           40.478            -74.286   
25%          6.000           -73.992           40.737            -73.992   
50%          8.500           -73.982           40.753            -73.981   
75%         12.500           -73.968           40.768            -73.965   
max        952.000           -73.700           41.361            -73.700   

       dropoff_latitude  passenger_count  
count       9778871.000      9778871.000  
mean             40.751            1.685  
std               0.031            1.308  
min              40.478            0.000  
25%              40.736            1.000  
50%              40.754            1.000  
75%              40.768            2.000  
max              41.358            6.000